Updating Json package readme (#1)
ericstj/runtime: src/libraries/System.Text.Json/src/PACKAGE.md

## About

> <!-- A description of the package and where one can find more documentation -->

Reviewer: What's the significance of XML remarks? Providing a template for other packages?

Author: Yes, this is the template that will remain.

Provides high-performance and low-allocating types that serialize objects to JavaScript Object Notation (JSON) text and deserialize JSON text to objects, with UTF-8 support built-in. Also provides types to read and write JSON text encoded as UTF-8, and to create an in-memory document object model (DOM), that is read-only, for random access of the JSON elements within a structured view of the data.

## How to Use

<!-- A compelling example on how to use this package with code, as well as any specific guidelines for when to use the package -->

> The System.Text.Json library is built-in as part of the shared framework in .NET Runtime. The package can be installed when you need to use it in other target frameworks.

Reviewer: ```suggestion
The System.Text.Json library is built-in as part of the shared framework in .NET Runtime. The package can be installed when you need to use the most recent version in older target frameworks.
```

WeatherForecast forecast = new (DateTimeOffset.Now, 26.6f, "Sunny");
var serialized = JsonSerializer.Serialize(forecast);

Console.WriteLine(serialized);
// {"Date":"2023-08-02T16:01:20.9025406+00:00","TemperatureCelsius":26.6,"Summary":"Sunny"}

var forecastDeserialized = JsonSerializer.Deserialize<WeatherForecast>(serialized);
Console.WriteLine(forecast == forecastDeserialized);
// True

public record WeatherForecast(DateTimeOffset Date, float TemperatureCelsius, string? Summary);
```

Serialization using the source generator:
```csharp
using System.Text.Json;
using System.Text.Json.Serialization;

WeatherForecast forecast = new (DateTimeOffset.Now, 26.6f, "Sunny");
var serialized = JsonSerializer.Serialize(forecast, SourceGenerationContext.Default.WeatherForecast);

Console.WriteLine(serialized);
// {"Date":"2023-08-02T16:01:20.9025406+00:00","TemperatureCelsius":26.6,"Summary":"Sunny"}

var forecastDeserialized = JsonSerializer.Deserialize<WeatherForecast>(serialized, SourceGenerationContext.Default.WeatherForecast);
Console.WriteLine(forecast == forecastDeserialized);
// True

public record WeatherForecast(DateTimeOffset Date, float TemperatureCelsius, string? Summary);

> [JsonSourceGenerationOptions(WriteIndented = true)]

Reviewer: The JSON output in the comments disagrees with the `WriteIndented` setting. Perhaps we could just remove `JsonSourceGenerationOptions` since it's for more advanced scenaria?

Author: ```suggestion
```
Sure.

[JsonSerializable(typeof(WeatherForecast))]
internal partial class SourceGenerationContext : JsonSerializerContext
{
}
```

Using the JSON DOM:
```csharp

using System;
using System.Text.Json;
using System.Text.Json.Nodes;

string jsonString =
@"{
  ""Date"": ""2019-08-01T00:00:00"",
  ""Temperature"": 25,
  ""Summary"": ""Hot"",
  ""DatesAvailable"": [
    ""2019-08-01T00:00:00"",
    ""2019-08-02T00:00:00""
  ],
  ""TemperatureRanges"": {
      ""Cold"": {
          ""High"": 20,
          ""Low"": -10
      },
      ""Hot"": {
          ""High"": 60,
          ""Low"": 20
      }
  }
}
";

JsonNode forecastNode = JsonNode.Parse(jsonString)!;

// Get value from a JsonNode.

> JsonNode temperatureNode = forecastNode!["Temperature"]!;

Reviewer: ```suggestion
JsonNode temperatureNode = forecastNode["Temperature"]!;
```

Author: All these came from the official conceptual docs - so we may want to follow up to get those fixed: https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-dom#use-jsonnode

Console.WriteLine($"Type={temperatureNode.GetType()}");
Console.WriteLine($"JSON={temperatureNode.ToJsonString()}");
//output:
//Type = System.Text.Json.Nodes.JsonValue`1[System.Text.Json.JsonElement]
//JSON = 25

// Get a typed value from a JsonNode.

> int temperatureInt = (int)forecastNode!["Temperature"]!;

Reviewer: ```suggestion
int temperatureInt = (int)forecastNode["Temperature"]!;
```

Console.WriteLine($"Value={temperatureInt}");
//output:
//Value=25

// Get a typed value from a JsonNode by using GetValue<T>.

> temperatureInt = forecastNode!["Temperature"]!.GetValue<int>();

Reviewer: ```suggestion
temperatureInt = forecastNode["Temperature"]!.GetValue<int>();
```

Console.WriteLine($"TemperatureInt={temperatureInt}");
//output:
//Value=25

// Get a JSON object from a JsonNode.

> JsonNode temperatureRanges = forecastNode!["TemperatureRanges"]!;

Reviewer: ```suggestion
JsonNode temperatureRanges = forecastNode["TemperatureRanges"]!;
```

Console.WriteLine($"Type={temperatureRanges.GetType()}");
Console.WriteLine($"JSON={temperatureRanges.ToJsonString()}");
//output:
//Type = System.Text.Json.Nodes.JsonObject
//JSON = { "Cold":{ "High":20,"Low":-10},"Hot":{ "High":60,"Low":20} }

// Get a JSON array from a JsonNode.

> JsonNode datesAvailable = forecastNode!["DatesAvailable"]!;

Reviewer: ```suggestion
JsonNode datesAvailable = forecastNode["DatesAvailable"]!;
```

Console.WriteLine($"Type={datesAvailable.GetType()}");
Console.WriteLine($"JSON={datesAvailable.ToJsonString()}");
//output:
//datesAvailable Type = System.Text.Json.Nodes.JsonArray
//datesAvailable JSON =["2019-08-01T00:00:00", "2019-08-02T00:00:00"]

// Get an array element value from a JsonArray.
JsonNode firstDateAvailable = datesAvailable[0]!;
Console.WriteLine($"Type={firstDateAvailable.GetType()}");
Console.WriteLine($"JSON={firstDateAvailable.ToJsonString()}");
//output:
//Type = System.Text.Json.Nodes.JsonValue`1[System.Text.Json.JsonElement]
//JSON = "2019-08-01T00:00:00"

// Get a typed value by chaining references.
int coldHighTemperature = (int)forecastNode["TemperatureRanges"]!["Cold"]!["High"]!;
Console.WriteLine($"TemperatureRanges.Cold.High={coldHighTemperature}");
//output:
//TemperatureRanges.Cold.High = 20

// Parse a JSON array

> var datesNode = JsonNode.Parse(@"[""2019-08-01T00:00:00"",""2019-08-02T00:00:00""]");

Reviewer: ```suggestion
JsonNode datesNode = JsonNode.Parse(@"[""2019-08-01T00:00:00"",""2019-08-02T00:00:00""]")!;
```

> JsonNode firstDate = datesNode![0]!.GetValue<DateTime>();

Reviewer: ```suggestion
JsonNode firstDate = datesNode[0]!.GetValue<DateTime>();
```

Console.WriteLine($"firstDate={ firstDate}");
//output:
//firstDate = "2019-08-01T00:00:00"
```

> Using the reader/writer

Reviewer: ```suggestion
Using the low-level JSON reader/writer types
```

```csharp
/*

>   SharpLab tools in Run mode:

Reviewer: Was this included accidentally?

Author: Yes you caught me! 😆

var jsonBytes = stream.ToArray();
string json = Encoding.UTF8.GetString(jsonBytes);
Console.WriteLine(json);
// {
//   "date": "2023-08-02T09:00:00+00:00"
//   "temp": 42
// }

var readerOptions = new JsonReaderOptions
{
    AllowTrailingCommas = true,
    CommentHandling = JsonCommentHandling.Skip
};
var reader = new Utf8JsonReader(jsonBytes, readerOptions);

while (reader.Read())
{
    Console.Write(reader.TokenType);

    switch (reader.TokenType)
    {
        case JsonTokenType.PropertyName:
        case JsonTokenType.String:
            {
                string? text = reader.GetString();
                Console.Write(" ");
                Console.Write(text);
                break;
            }

        case JsonTokenType.Number:
            {
                int intValue = reader.GetInt32();
                Console.Write(" ");
                Console.Write(intValue);
                break;
            }

            // Other token types elided for brevity
    }
    Console.WriteLine();
}
// StartObject
// PropertyName date
// String 2023-08-02T09:00:00+00:00
// PropertyName temp
// Number 42
// EndObject
```

## Key Features

<!-- The key features of this package -->

> TODO

Reviewer: My impression is that a "Key Features" section would essentially be a recap of the APIs already showcased in the "How to Use" section. Perhaps it should go before the code samples?

Author: Yeah, that would be fine.  I need your help with a bulleted list here.

## Main Types

<!-- The main types provided in this library -->

The main types provided by this library are:

> * `System.Text.Json.JsonSerializer`

Reviewer: `JsonSerializerOptions` is fairly pervasive and important. Should we include it here or does it get covered by the "JsonSerializer" umbrella?

Author: Yep - please add whichever types you think folks will think of as entry point types.  No need to ask - just add them.

* `System.Text.Json.JsonSerializer`
* `System.Text.Json.JsonDocument`
* `System.Text.Json.JsonElement`
* `System.Text.Json.Nodes.JsonNode`
* `System.Text.Json.Utf8JsonWriter`
* `System.Text.Json.Utf8JsonReader`

## Addtional Documentation

* [Conceptual documentation](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/overview)
* [API documentation](https://learn.microsoft.com/en-us/dotnet/api/system.text.json)

## Related Packages

<!-- The related packages associated with this package -->

* Lightweight data formats abstraction: [System.Memory.Data](https://www.nuget.org/packages/System.Memory.Data/)
* Serialization of HttpContent: [System.Net.Http.Json](https://www.nuget.org/packages/System.Net.Http.Json/)

## Feedback & Contributing

> For more information, see the documentation:

Reviewer: I recall that when we originally added this document we deliberately avoided inlining documentation content to avoid creating a second source of truth that would eventually go out of sync.

Author: Correct, so what's going here is mostly links and some simple samples.  Add any additional links you'd like, or reduce the samples if you feel there is too much duplication.

The links that show as removed here  are covered by the section `Additional Documentation` above.
